### PR TITLE
LPS-125607 Fix select any element from the highlighted ones and delete click target

### PR DESCRIPTION
**Pull request by dgarciasarai**
**Description**

When we create an AB Test with Click goal, we want to let the user enter the element ID (click goal target). Otherwise, the user can always select the click goal target from the UI (with the highlighted ones).

**Impedibug**

With this commit https://github.com/liferay-frontend/liferay-portal/pull/650/commits/fc06def393307e7b3a1d852f9373207f99fd6ff3 (already merged in master) we lose two functionalities:

- Select the click goal target with the UI (from the highlighted ones)
- Remove the selected click goal target (to be empty again)

**Screenshot**

LPS-125607

### CODE REVIEW

**Comment by dgarciasarai** [comment]
Hey @liferay-frontend, can you give me feedback? Specially @bryceosterhaus that was the one helping the team with this feature :) Thanks!

File: modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js [hunk 1]
```diff
@@ -370,25 +370,20 @@ function OverlayContainer({allowEdit, root}) {
...
 
-			mousedownRef.current = false;
+			clickRef.current = !clickRef.current;
 		},
 		[dispatch]
 	);
 
-	useEventListener('mousedown', handleMouseDown, false, document);
-	useEventListener('mouseup', handleMouseUp, false, document);
+	useEventListener('click', handleClick, false, document);
```
Comment (bryceosterhaus) [inline modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js:386]: So from what I remember, I think changing this to a `click` event will cause a blur event and will cause the input to lose focus and then the element is selected but not highlighted.

File: modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js [hunk 2]
```diff
@@ -370,25 +370,20 @@ function OverlayContainer({allowEdit, root}) {
...
 
-	const handleMouseDown = useCallback(() => {
-		mousedownRef.current = true;
-	}, []);
-
-	const handleMouseUp = useCallback(
+	const handleClick = useCallback(
 		(event) => {
-			if (mousedownRef.current === true) {
+			if (clickRef.current === true) {
```
Comment (bryceosterhaus) [inline modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js:375]: If you move this event from `mousedown` to `click`, this check would likely cause some issues, as the first click on the page would have to happen first before this ref is set to true.

File: modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js (hunk 1 above)
Comment (bryceosterhaus) [reply modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js:386]: To test this bug, remove the `if ()` wrapper in the event handler and see how clicking outside of the input and hitting enter in the input have two different results.

Comment (dgarciasarai) [reply modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js:386]: > So from what I remember, I think changing this to a click event will cause a blur event and will cause the input to lose focus and then the element is selected but not highlighted.

Yes, that happened with click event and no ref. In this case I used click event and ref. I mixed both concepts. I don't know if that is correct or not. With this change, if the user enters an ID in the input and hits enter and then wants to change it using the highlighted ones, the first click would lose the focus in the input and the second will be 1) cancel edition if the user clicks in a non-clickable element; 2) open the clickable element to select as target (popover + topper box view).

**Comment by dgarciasarai** [comment]
Hey @bryceosterhaus, thank you for your feedback. I know that this is not a good solution and maybe cases other problems in the functionality. I am not very familiar with refs :( Do you have another idea about how to solve this impedibug? I'm going to try to use `mousedown` event but right now, the first mousedown event, the user enters into the "active" mode (with the highlighted elements in the UI), the second one, this elements "disappear" (are not highlighted) and you cannot select it. Also, if you have a click goal target, you cannot remove it (the close icon doesn't work for the same reason).

Thanks!

**Comment by dgarciasarai** [comment]
Feel free to update the code/send any other solution you found! :)

File: modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js [hunk 3]
```diff
@@ -376,9 +378,11 @@ function OverlayContainer({allowEdit, root}) {
...
 	const handleMouseUp = useCallback(
 		(event) => {
-			if (mousedownRef.current === true) {
-				event.preventDefault();
-				stopImmediatePropagation(event);
+			const overlayTarget = event.target.closest(
+				`.${OVERLAY_TARGET_CLASS}`
+			);
+
+			if (mousedownRef.current === true && !overlayTarget) {
```
Comment (bryceosterhaus) [inline modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/ClickGoalPicker.es.js:385]: The issue here was that we were calling `deactivate` when the user was clicking on an overlay target. By adding this logic, we make sure to not deactivate if its an overlay target. You should not be able to select a new target and also remove an existing target

**Review by bryceosterhaus — COMMENTED** [review]
@dgarciasarai I think I figured out the issue. The events were clobbering each other and deactivating the overlay. Let me know what you think or if I missed anything

BTW, before forwarding this, we'll want to remove the first 2 commits.

**Comment by dgarciasarai** [comment]
Thank you so much @bryceosterhaus, now it's working as expected! You're a hero! 🚀

**Comment by dgarciasarai** [comment]
ci:forward

**Pull request closed without merging**